### ISSUE TITLE
C99 issues

### ISSUE THREAD

**Issue by embeddedt** [lvgl/lvgl#890]
@kisvegabor There is a portability issue I'm seeing while testing how LittlevGL compiles with very strict GCC warnings. ISO C99 doesn't support unnamed structs/unions.

Example:

https://github.com/littlevgl/lvgl/blob/a374baa37602288081a5ad770a875089a95acfbf/lv_hal/lv_hal_indev.h#L73-L91

The `union` itself is unnamed and contains two unnamed structs.

**Comment by C47D** [comment]
Anonymous enums or structs are available since C11 [1].

[1] https://www.geeksforgeeks.org/g-fact-38-anonymous-union-and-structure/

**Comment by embeddedt** [comment]
@C47D That's true, but we're supposed to have C99 compatibility.

**Comment by kisvegabor** [comment]
Thank you for pointing this out. 

These structs will be updated by the muli-display update. So let's postpone it until it's merged. (probably next week)

Have you found other unnamed structs or enums too?

**Comment by embeddedt** [comment]
@kisvegabor I think there may have been others, but this one was generating a lot of the warnings, since it's included by most of the files. I will report other ones if I see them.

**Comment by kisvegabor** [comment]
I named these struct and unions in `dev-6.0`.

**Comment by embeddedt** [comment]
@kisvegabor Great, I'll try it again as soon as possible.

**Comment by embeddedt** [comment]
@kisvegabor These unions are still unnamed, as well as the ones in the `lv_colorX_t` types in `lv_misc/lv_color.h`:

https://github.com/littlevgl/lvgl/blob/551eadc1d5fc6c20bab9c88ff7dcd1d5f0ae37e4/lv_hal/lv_hal_indev.h#L59-L69

https://github.com/littlevgl/lvgl/blob/551eadc1d5fc6c20bab9c88ff7dcd1d5f0ae37e4/lv_hal/lv_hal_indev.h#L125-L136

Additionally, a new problem was introduced with multi-display support.
```
./lvgl/lv_conf_checker.h:198:16: warning: redefinition of typedef 'lv_disp_drv_user_data_t' [-Wpedantic]
```
Redefining types is not allowed in C, nor is there a facility to check whether the type is already defined.

**Comment by kisvegabor** [comment]
Thank you!

I've updated the colors. I'll do the rest soon.

**Comment by kisvegabor** [comment]
@embeddedt 
Do you have an idea about what to do with the `typedefs`? It can't be checked by `lv_conf_checker.h` if a `typedef` exists or not. 

I can't imagine a better solution than simply not checking `typedef`s in `lv_conf_checker.h`. What do you think?

**Comment by embeddedt** [comment]
@kisvegabor I agree. Let's remove the typedefs from `lv_conf_checker.h`.

**Comment by kisvegabor** [comment]
Ok, I'll do this way.
I fixed these issues in the `event` branch which will be merged soon into `dev-6.0`

**Comment by beibean** [comment]
I am having problems when using next functions:

```C
uint8_t (*lv_txt_encoded_size)(const char *) =                  lv_txt_utf8_size;
uint32_t (*lv_txt_unicode_to_encoded)(uint32_t) =               lv_txt_unicode_to_utf8;
uint32_t (*lv_txt_encoded_conv_wc)(uint32_t) =                  lv_txt_utf8_conv_wc;
uint32_t (*lv_txt_encoded_next)(const char *, uint32_t *) =    lv_txt_utf8_next;
uint32_t (*lv_txt_encoded_prev)(const char *, uint32_t *) =   lv_txt_utf8_prev;
uint32_t (*lv_txt_encoded_get_byte_id)(const char *, uint32_t) = lv_txt_utf8_get_byte_id;
uint32_t (*lv_encoded_get_char_id)(const char *, uint32_t) =   lv_txt_utf8_get_char_id;
uint32_t (*lv_txt_get_encoded_length)(const char *) =          lv_txt_utf8_get_length;
```

The question here is, Is this type of initialization legal in C?

What I thought is that a global variable could only be initialized to a constant or a literal since it is done in compiling time.

When assigning a function pointer, like it's the case, its address might not be allocated yet.

**Edit**
I have the same problem when using fonts within the device.

First declared:
```C
LV_FONT_DECLARE(lv_font_dejavu_20)
```

And then defined:
```C
lv_font_t lv_font_dejavu_20 = {
    .unicode_first = 32,    /*First Unicode letter in this font*/
    .unicode_last = 126,    /*Last Unicode letter in this font*/
    .h_px = 20,             /*Font height in pixels*/
    .glyph_bitmap = lv_font_dejavu_20_glyph_bitmap, /*Bitmap of glyphs*/
    .glyph_dsc = lv_font_dejavu_20_glyph_dsc,       /*Description of glyphs*/
    .glyph_cnt = 95,            /*Number of glyphs in the font*/
    .unicode_list = NULL,   /*Every character in the font from 'unicode_first' to 'unicode_last'*/
    .get_bitmap = lv_font_get_bitmap_continuous,    /*Function pointer to get glyph's bitmap*/
    .get_width = lv_font_get_width_continuous,  /*Function pointer to get glyph's width*/
#if USE_LV_FONT_DEJAVU_20 == 1
    .bpp = 1,               /*Bit per pixel*/
#elif USE_LV_FONT_DEJAVU_20 == 2
    .bpp = 2,               /*Bit per pixel*/
#elif USE_LV_FONT_DEJAVU_20 == 4
    .bpp = 4,               /*Bit per pixel*/
#elif USE_LV_FONT_DEJAVU_20 == 8
    .bpp = 8,               /*Bit per pixel*/
#endif
    .monospace = 0,
    .next_page = NULL,      /*Pointer to a font extension*/
};
```

So, it seems that what my device doesn't like is declaring a global variable in one site and giving value to it in other site (compiling time).

I don't have that problem when declaring a variable and give a value straight away. As for example:

```C
static void (*px_fp)(lv_coord_t x, lv_coord_t y, const lv_area_t * mask, lv_color_t color, lv_opa_t opa) = lv_vpx;
static void (*fill_fp)(const lv_area_t * coords, const lv_area_t * mask, lv_color_t color, lv_opa_t opa) =  lv_vfill;
static void (*letter_fp)(const lv_point_t * pos_p, const lv_area_t * mask, const lv_font_t * font_p, uint32_t letter, lv_color_t color, lv_opa_t opa) = lv_vletter;
```

Do you know if it is any problem with my compilation/micro or is related with C99/C11 standard?

**Comment by embeddedt** [comment]
@beibean What is the exact error you are getting from the compiler?

**Comment by embeddedt** [comment]
@beibean In that case it sounds like an issue with your device specifically. I think that there is no undefined behavior used here.

**Comment by beibean** [comment]
@embeddedt Ok, thanks! I will use proper initialization functions myself.

**Comment by beibean** [comment]
@embeddedt What about this define within lv_conf?

```C
#define LV_COMPILER_NON_CONST_INIT_SUPPORTED 1  /* 1: Initialization with non constant values are supported */
```

**Comment by embeddedt** [comment]
@beibean You can try configuring that according to how your compiler setup works.

**Comment by kisvegabor** [comment]
What is the difference between:
```c
/*Working*/
static void (*px_fp)(lv_coord_t x, lv_coord_t y, const lv_area_t * mask, lv_color_t color, lv_opa_t opa) = lv_vpx;
```

and 

```c
/*Not working*/
uint8_t (*lv_txt_encoded_size)(const char *) =     lv_txt_utf8_size;
```

It seems the same to me: assign a function pointer to a global function pointer variable.

**Comment by beibean** [comment]
> You can try configuring that according to how your compiler setup works.

There are not references in the code to that define. So no changes will be done.

>It seems the same to me: assign a function pointer to a global function pointer variable.

There is a big difference:

Static variable is initialized when declared while extern variable is not.

Just for you to know, this variable within `lv_draw_aa_get_opa` did not initialize properly neither:

```C
    static const lv_opa_t * seg_map[] = {seg1, seg2, seg3, seg4,
                                         seg5, seg6, seg7, seg8
                                        };
```

**Comment by kisvegabor** [comment]
> Static variable is initialized when declared while extern variable is not.

AFAIK all `static` and `global` variables are initialized in the `startup` code.

So these are equivalent for the compiler:
```c
uint8_t global_x = 1;
static uint8_t static global_x = 1;

int main(void) {
  static uint8_t static_local = 1;

}

All these variables are mapped to specific RAM addresses and initialized before `main` is called

```


Anyway, I believe it works correctly because in the past few years nobody reported any issue about it.

**Comment by beibean** [comment]
> AFAIK all static and global variables are initialized in the startup code.

That is what I thought, but somehow my compiler treats them in a different manner.

> Anyway, I believe it works correctly because in the past few years nobody reported any issue about it.

No bother, I have already created my own init functions.

I do not have much time these days, but I will investigate more deeply this issue and let you know what is happening with my compiler, just for the case someone could have the same matter.

**Comment by kisvegabor** [comment]
@beibean I'm also curious what can cause this issue.

@embeddedt I merged a lot of things to `dev-6.0` today. It also fixes the issues mentioned above. Can you check what is still missing?

**Comment by embeddedt** [comment]
@kisvegabor There were a few remaining problems; I fixed them in #948.

**Comment by kisvegabor** [comment]
Merged, thank you very much!

As the main goal of this issue is reach I close it.
